addons/hw_drivers/websocket_client.py

```diff
@@ -106,3 +54,75 @@ def run(self):
                 _logger.exception("An unexpected exception happened when running the websocket")
             _logger.debug('websocket will try to restart in 10 seconds')
             time.sleep(10)
+
+    @staticmethod
+    def send_to_controller(params):
+        """Confirm the operation's completion by sending
+        a response back to the Odoo server.
+
+        :param params: The parameters to send back to the server
+        """
+        server_url = helpers.get_odoo_server_url() + "/iot/box/send_websocket"
+        try:
+            response = requests.post(server_url, json={'params': params}, timeout=5)
+            response.raise_for_status()
+        except requests.exceptions.RequestException:
+            _logger.exception('Could not reach confirmation status URL: %s', server_url)
+
+    def on_open(self, ws):
+        """When the client is set up, this function sends a message
+        to subscribe to the iot websocket channel.
+
+        :param ws: The websocket client
+        """
+        ws.send(
+            json.dumps({
+                'event_name': 'subscribe',
+                'data': {
+                    'channels': [self.iot_channel],
+                    'last': 0,
+                    'mac_address': helpers.get_mac_address()
+                }
+            })
+        )
+
+    def on_message(self, _ws, messages):
+        """Synchronously handle messages received by the websocket.
+
+        :param _ws: The websocket client
+        :param messages: The message list received by the websocket
+        """
+        for message in json.loads(messages):
+            message = message['message']
+            _logger.debug("websocket received a message: %s", pprint.pformat(message))
+
+            msg_type = message['type']
+            if msg_type == 'operation_confirmation':
+                return
+            if msg_type != 'iot_action':
+                _logger.warning("Message type not supported: %s", msg_type)
+                return
+
+            payload = message['payload']
+            iot_box_ip = payload.get('iot_box_ip')
+            if iot_box_ip != helpers.get_domain():
+                # likely intended as IoT Boxes share the same channel
+                _logger.debug("Message ignored due to different iot box IP: %s", iot_box_ip)
+                return
```

Comment:
to discuss the replacing of the mac address by the domain

Reply:
Currently:
 
- websocket uses MAC,
- longpolling uses IP.

Arguments to use IP Address as a unique identifier:
- We use one websocket channel by company, reducing the risk of having duplicate IPs on the network,
- if we had duplicate IPs, we would already encounter issues with generating an HTTPs certificate for the IoT Box (as it's format is `0-0-0-0.domain.tld`: it would result in duplicated subdomains),
- We would also encounter issues using longpolling, contacting two IoT Boxes for the same request.

In my opinion, the IP address ALREADY is a unique identifier for IoT Boxes.